packages/loader/container-loader/src/runtimeCompatManager.ts

```diff
@@ -0,0 +1,68 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+	LayerCompatibilityManager,
+	type ICompatibilityDetails,
+} from "@fluid-internal/client-utils";
+import type { ICriticalContainerError } from "@fluidframework/container-definitions";
+import { UsageError } from "@fluidframework/telemetry-utils/internal";
+
+import { pkgVersion } from "./packageVersion.js";
+
+/**
+ * This class manages the state and validation to ensure that the Loader layer is compatible with the Runtime layer.
+ * @internal
+ */
+export class RuntimeLayerCompatManager extends LayerCompatibilityManager {
+	// Minimum generation that Runtime must be at to be compatible with Loader.
+	public readonly runtimeMinSupportedGeneration = 1;
+	// The features that the Runtime must support to be compatible with Loader.
+	public readonly runtimeRequiredFeatures: string[] = [];
+
+	public constructor(private readonly disposeFn: (error?: ICriticalContainerError) => void) {
+		super({
+			// The current package version of the Loader layer.
+			pkgVersion,
+			// The current generation of the Loader layer.
+			generation: 1,
+			// The features supported by the Loader layer across Loader <-> Runtime boundary.
+			supportedFeatures: new Set(),
+		});
+	}
+
+	/**
+	 * Validates that the Runtime layer is compatible with the Loader.
+	 */
+	public validateCompatibility(
+		maybeRuntimeCompatDetails: ICompatibilityDetails | undefined,
+	): void {
+		// For backwards compatibility - until required features are added, Runtime is considered
+		// to be compatible. This is to allow Loader to work with existing Runtimes. Once we start
+		// enforcing layer compatibility, we will add required features or remove this check.
+		if (this.runtimeRequiredFeatures.length === 0) {
+			return;
+		}
+
+		const layerCheckResult = super.checkCompatibility(
+			this.runtimeMinSupportedGeneration,
+			this.runtimeRequiredFeatures,
+			maybeRuntimeCompatDetails,
+		);
+		if (!layerCheckResult.isCompatible) {
```

Comment:
Close and throw only if there are unsupported features. If only generation is incompatible, log an error / warning.